consensus/tendermint/propose_test.go

```diff
@@ -0,0 +1,536 @@
+
+		assert.Equal(t, propose, algo.state.step)
+		assert.Equal(t, expectedHeight, algo.state.height)
+		assert.Equal(t, rPrime, algo.state.round)
+	})
+
+	t.Run("Line 55 (Prevote): Start round r' when f+1 future round messages are received from round r'", func(t *testing.T) {
+		listeners, broadcasters := testListenersAndBroadcasters()
+		app, chain, vals := newApp(), newChain(), newVals()
+
+		vals.addValidator(*val2)
+		vals.addValidator(*val3)
+		vals.addValidator(*val4)
+		vals.addValidator(*nodeAddr)
+
+		algo := New[value, felt.Felt, felt.Felt](*nodeAddr, app, chain, vals, listeners, broadcasters, tm, tm, tm)
+
+		expectedHeight := uint(0)
+		rPrime := uint(4)
+		round4Value := value(10)
+		val2Proposal := Proposal[value, felt.Felt, felt.Felt]{
+			Height:     expectedHeight,
+			Round:      rPrime,
+			ValidRound: nil,
+			Value:      &round4Value,
+			Sender:     *val2,
+		}
+
+		val3Prevote := Prevote[felt.Felt, felt.Felt]{
+			Vote: Vote[felt.Felt, felt.Felt]{
+				Height: expectedHeight,
+				Round:  rPrime,
+				ID:     utils.Ptr(round4Value.Hash()),
+				Sender: *val3,
+			},
+		}
+
+		algo.messages.addProposal(val2Proposal)
+		prevoteListener := listeners.PrevoteListener.(*senderAndReceiver[Prevote[felt.Felt, felt.Felt], value,
+			felt.Felt, felt.Felt])
+		prevoteListener.send(val3Prevote)
+
+		algo.Start()
+		time.Sleep(1 * time.Millisecond)
+		algo.Stop()
+
+		assert.Equal(t, 1, len(algo.messages.proposals[expectedHeight][rPrime][*val2]))
+		assert.Equal(t, 1, len(algo.messages.prevotes[expectedHeight][rPrime][*val3]))
+		assert.Equal(t, val3Prevote, algo.messages.prevotes[expectedHeight][rPrime][*val3][0])
+
+		assert.Equal(t, propose, algo.state.step)
+		assert.Equal(t, expectedHeight, algo.state.height)
+		assert.Equal(t, rPrime, algo.state.round)
+	})
+
+	t.Run("Line 55 (Precommit): Start round r' when f+1 future round messages are received from round r'", func(t *testing.T) {
+		listeners, broadcasters := testListenersAndBroadcasters()
+		app, chain, vals := newApp(), newChain(), newVals()
+
+		vals.addValidator(*val2)
+		vals.addValidator(*val3)
+		vals.addValidator(*val4)
+		vals.addValidator(*nodeAddr)
+
+		algo := New[value, felt.Felt, felt.Felt](*nodeAddr, app, chain, vals, listeners, broadcasters, tm, tm, tm)
+
+		expectedHeight := uint(0)
+		rPrime := uint(4)
+		round4Value := value(10)
+		val2Precommit := Precommit[felt.Felt, felt.Felt]{
+			Vote: Vote[felt.Felt, felt.Felt]{
+				Height: expectedHeight,
+				Round:  rPrime,
+				ID:     utils.Ptr(round4Value.Hash()),
+				Sender: *val2,
+			},
+		}
+
+		val3Prevote := Prevote[felt.Felt, felt.Felt]{
+			Vote: Vote[felt.Felt, felt.Felt]{
+				Height: expectedHeight,
+				Round:  rPrime,
+				ID:     utils.Ptr(round4Value.Hash()),
+				Sender: *val3,
+			},
+		}
+
+		algo.messages.addPrevote(val3Prevote)
+		prevoteListener := listeners.PrecommitListener.(*senderAndReceiver[Precommit[felt.Felt, felt.Felt], value,
```

Comment:
typo. Also, do you think we could collapse the above three tests into a single test? It seems most of the code is repeated

Reply:
I was trying to make the tests independent. If we start sharing variables, it becomes difficult to debug because of side effects. I have variables that don't change at the top of the test.